[dwrite] GCC/mingw/msys2 compatibility

ebraminio opened the pull request:
It now compiles on GCC also but msys2 needs to be patched like this https://github.com/Alexpux/mingw-w64/pull/5

behdad commented on src/hb-directwrite.cc:
```diff
@@ -919,7 +924,7 @@ hb_directwrite_shape_experimental_width(hb_font_t          *font,
   float               width)
 {
-  static char *shapers = "directwrite";
+  static char *shapers = (char *) "directwrite";
```
This is wrong. You should just declare "static const char *...".

ebraminio replied:
Oh, thanks.

behdad commented:
Can you explain what this change does?

ebraminio commented:
> Can you explain what this change does?

hb-directwrite was never compiled on GCC, I fixed headers missing of mingw64 project and applied fixes needed to get it working on GCC/msys2.

ebraminio commented:
* Replaced new/delete with plain malloc to get it working on autotools.
* Replaced the places false meant `nullptr` actually but MSVC didn't complain.
* Fixed the places that compiler raised warnings, of course using const was more appropriate and I just wanted to make it happy, and I wasn't enlightened with `const` as the time originally writing this! :)

There are more improvement needed for hb-directwrite of course now that I see but as hb-directwrite wasn't considered that important from the beginning I just wanted to fix it on GCC and no more.

behdad commented:
I see.  Thanks. That's a lot more helpful :).
